### PR TITLE
add DispatcherOP

### PR DESCRIPTION
**Pull request by njzjz**

### CODE REVIEW

File: dpgen/ops/dispatcherop.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+    @property
+    def backward_common_files(self) -> List[str]:
+        return []
+    
+    @property
+    def task_paths(self) -> List[Path]:
+        if self.status is not Status.EXECUTED:
+            raise RuntimeError
+        return self.work_path.glob(self.task_pattern)
```
Comment (amcadmus) [inline dpgen/ops/dispatcherop.py]: We may not want to use pattern to find tasks, but accept a set of Path objects. For example the "train_dirs" as output of `PrepDPTrain`

https://github.com/amcadmus/dpgen/blob/05a74613449ae67867328d1f67d949a0a99fe45f/dpgen/ops/prep_dp_train.py#L122-L127

Comment (njzjz) [reply dpgen/ops/dispatcherop.py]: I haven't understood it...

Comment (amcadmus) [reply dpgen/ops/dispatcherop.py]: I suggest that we d not use task_pattern, but explicitly pass the Set of tasks.

Comment (njzjz) [reply dpgen/ops/dispatcherop.py]: `self.numb_models` is set by users, but, do users know really know how many tasks there will be before running the OP?

Comment (amcadmus) [reply dpgen/ops/dispatcherop.py]: Yes, the working path of training tasks are output by `PrepDPTrain`, thus the dpgen knows which training task to run.

File: dpgen/ops/dispatcherop.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            machine=self.machine,
+            resources=self.resources,
+            task_list=tasks,
+            forward_common_files=self.path_to_str(self.forward_common_files),
+            backward_common_files=self.path_to_str(self.backward_common_files),
+        )
+        submission.run_submission()
+        return op_in
+
+    @classmethod
```
Comment (amcadmus) [inline dpgen/ops/dispatcherop.py:115]: Shouldn't it be a static method?

File: dpgen/ops/dispatcherop.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        return "log"
+    
+    @property
+    @abstractmethod
+    def forward_files(self) -> Set[Path]:
+        """List[str]: relative path to task path"""
+
+    @property
+    @abstractmethod
+    def backward_files(self) -> Set[Path]:
```
Comment (amcadmus) [inline dpgen/ops/dispatcherop.py:59]: abstractmethod does not have `self` as input

File: dpgen/ops/dispatcherop.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        DPDispatcher machine parameters. See also DPDispatcher doc.
+    resources : dict
+        DPDispatcher machine parameters. See also DPDispatcher doc.
+    command : str
+        path or command to a certain program on remote
+    """
+    @OP.set_status(status = Status.INITED)
+    def __init__(
+            self,
+            context: IterationContext,
```
Comment (amcadmus) [inline dpgen/ops/dispatcherop.py]: It seems that `Context` should be enough

File: dpgen/ops/dispatcherop.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+            machine: dict,
+            resources: dict,
```
Comment (amcadmus) [inline dpgen/ops/dispatcherop.py]: How about input Machine and Resources objects?

File: dpgen/ops/dispatcherop.py (hunk 3 above)
Comment (njzjz) [reply dpgen/ops/dispatcherop.py:59]: Do you mean `staticmethod` does not have `self` as input?

File: dpgen/ops/dispatcherop.py (hunk 5 above)
Comment (njzjz) [reply dpgen/ops/dispatcherop.py]: That's okay, but we also need to convert dict to object outside the OP

File: dpgen/ops/dispatcherop.py (hunk 3 above)
Comment (amcadmus) [reply dpgen/ops/dispatcherop.py:59]: yes

Comment (njzjz) [reply dpgen/ops/dispatcherop.py:59]: but this is `abstractmethod` and not `staticmethod`...

**Review by amcadmus — APPROVED** [review]

**Pull request merged**